AWSClientRuntime/Sources/Middlewares/Sha256TreeHashMiddleware.swift

```diff
@@ -0,0 +1,118 @@
+                      return next.handle(context: context, input: input)
+                  }
+                  let (linearHash, treeHash) = computeHashes(bytes: streamBytes)
+                  if let treeHash = treeHash {
+                      input.withHeader(name: sha256TreeHashHeaderName, value: treeHash)
+                  }
+                  if let linearHash = linearHash {
+                      input.withHeader(name: sha256HashHeaderName, value: linearHash)
+                  }
+                  
+                  return next.handle(context: context, input: input)
+                  
+              case .empty, .none:
+                  return next.handle(context: context, input: input)
+              }
+              
+              return next.handle(context: context, input: input)
+          }
+    
+    /// Computes the tree-hash and linear hash of a `ByteBuffer`.
+    /// See http://docs.aws.amazon.com/amazonglacier/latest/dev/checksum-calculations.html for more information.
+    private func computeHashes(bytes: ByteBuffer) -> (String?, String?) {
+        let bufferSize = 1024 * 1024
+        var hashes = [[UInt8]]()
+        
+        while true {
+            var oneMbTempBuffer = ByteBuffer(size: bufferSize)
+            let bytesRead = bytes.readIntoBuffer(buffer: &oneMbTempBuffer)
+            if bytesRead == 0 {
+                break
+            }
+            let hash = oneMbTempBuffer.sha256()
+            hashes.append(hash.toByteArray())
+        }
+
+        return (bytes.base64EncodedSha256(), computeTreeHash(hashes: hashes))
+    }
+    
+    /// Builds a tree hash root node given a slice of hashes. Glacier tree hash to be derived from SHA256 hashes of 1MBm chunks of the data.
+    /// See http://docs.aws.amazon.com/amazonglacier/latest/dev/checksum-calculations.html for more information.
+    private func computeTreeHash(hashes: [[UInt8]]) -> String? {
+        guard !hashes.isEmpty else {
+            return nil
+        }
+        let hashCount = hashes.count
+        
+        if hashCount == 1 {
+            let byteArray = hashes[0]
+            let byteBuffer = ByteBuffer(bytes: byteArray)
+            return byteBuffer.toData().base64EncodedString()
+        }
+        var tempHashes = [[UInt8]]()
+        for index in stride(from: 0, to: hashCount, by: 2) {
+            if index + 1 <= hashCount - 1 {
+                var tempHashArray = [UInt8]()
+                tempHashArray.append(contentsOf: hashes[index])
+                tempHashArray.append(contentsOf: hashes[index + 1])
+                let tempByteBuffer = ByteBuffer(bytes: tempHashArray)
+                
+                let tempSha256 = tempByteBuffer.sha256()
+                tempHashes.append(tempSha256.toByteArray())
+                
+            } else {
+                tempHashes.append(hashes[index])
+            }
+        }
+         let byteBuf = ByteBuffer(bytes: tempHashes[0])
```

Comment:
1.  spacing seems off here
2.  This seems off, because in the last for loop we did a bunch of work to append a bunch of data to tempHashes, and in this line, we are only using the first byte.  Something seems off.

Reply:
1. corrected
2. we only need the root of the tree we have created. the root is at the zeroth index of tempHashes which is an array of hashes